python/cugraph/cugraph/gnn/data_loading/dist_sampler.py

```diff
@@ -0,0 +1,349 @@
+                : len(current_seeds)
+            ]
+
+            minibatch_dict = self.sample_batches(
+                seeds=current_seeds,
+                batch_ids=current_batches,
+                random_state=random_state,
+            )
+            self.__writer.write_minibatches(minibatch_dict)
+
+    @property
+    def is_multi_gpu(self):
+        return isinstance(self.__graph, pylibcugraph.MGGraph)
+
+    @property
+    def _local_seeds_per_call(self):
+        return self.__local_seeds_per_call
+
+    @property
+    def _graph(self):
+        return self.__graph
+
+
+class UniformNeighborSampler(DistSampler):
+    def __init__(
+        self,
+        graph: Union[pylibcugraph.SGGraph, pylibcugraph.MGGraph],
+        writer: DistSampleWriter,
+        *,
+        local_seeds_per_call: int = 32768,
+        fanout: List[int] = [-1],
+        prior_sources_behavior: str = "exclude",
+        deduplicate_sources: bool = True,
+        compression: str = "COO",
+        compress_per_hop: bool = False,
+        with_replacement: bool = False,
+    ):
+        super().__init__(graph, writer, local_seeds_per_call=local_seeds_per_call)
+        self.__fanout = fanout
+        self.__prior_sources_behavior = prior_sources_behavior
+        self.__deduplicate_sources = deduplicate_sources
+        self.__compress_per_hop = compress_per_hop
+        self.__compression = compression
+        self.__with_replacement = with_replacement
+
+    def sample_batches(
+        self, seeds: TensorType, batch_ids: TensorType, random_state: int = 0
+    ):
+        # FIXME allow skipping of the synchronization logic with a boolean
+        # flag that assumes all ranks have the same number of batches.
+        if self.is_multi_gpu:
+            rank = torch.distributed.get_rank()
+            handle = pylibcugraph.ResourceHandle(
+                cugraph_comms_get_raft_handle().getHandle()
+            )
+
+            local_label_list = torch.unique(batch_ids)
+            local_label_to_output_comm_rank = torch.full(
+                (len(local_label_list),), rank, dtype=torch.int32, device="cuda"
+            )
+
+            num_batches = torch.tensor(
+                [len(local_label_list)], device="cuda", dtype=torch.int64
+            )
+            torch.distributed.all_reduce(num_batches, op=torch.distributed.ReduceOp.SUM)
+
+            label_list = torch.empty((num_batches,), device="cuda", dtype=torch.int32)
+            w1 = torch.distributed.all_gather_into_tensor(
+                label_list, local_label_list, async_op=True
+            )
+
+            label_to_output_comm_rank = torch.empty(
+                (num_batches,), device="cuda", dtype=torch.int32
+            )
+            w2 = torch.distributed.all_gather_into_tensor(
+                label_to_output_comm_rank,
+                local_label_to_output_comm_rank,
+                async_op=True,
+            )
+
+            w1.wait()
+            w2.wait()
+
```

Comment:
Should we have a synchronization call here ? Given above are multi-node operations , this will probably lag at different places and the  `pylibcugraph.uniform_neighbor_sample` can probably hang when there is too much time b/w two calls . 

Or at-least it used to, will let @jnke2016 confirm about this as he has tested this the most.

Reply:
Doesn't that only apply if this is launched from a dask worker, which can die after a certain time?  I thought there is an implicit synchronization that occurs in the C++ code right before the call returns.  All the calls should end at roughly the same time.

Reply:
> which can die after a certain time? 

I think the scenario you are referring to timing out after hanging (which is actually a better practice for distributed frameworks IMO  and we should follow)  

>is an implicit synchronization that occurs in the C++ code right before the call returns. 
I am referring to before the `pylibcugraph.uniform_neighbor_sample`, will let @jnke2016 confirm about his experiments

Reply:
@VibhuJawa are you referring to the Jaccard/Two_hop_neighbor hang which transiently occurred if one or more processes are taking longer to return?

Reply:
Yes, I think that is an issue which should be true for other algos too unless i am mistaken.

Reply:
Right, the workaround I had was using `client.map()` with iterables (like you did a while ago experimentaly for neighborhood sampling).

Reply:
So what's the conclusion here then?  Do I need to add a delay somewhere?  Or is this way of calling sampling acceptable?

Reply:
But this workflow is quiet different than our MG algos as there is no dask. If the `torch` `all_gather` call is similar to `MPI/NCCL` then I don't think there is an issue. In fact, I had a C test adding a long delay to one or more ranks from the communicator prior to a blocking collective call and the other ranks simply waited and resumed afterward once the delayed rank were ready.

Reply:
`all_gather_into_tensor` here is a NCCL call

Reply:
Ok then I think this is fine.

Reply:
>I had a C test adding a long delay to one or more ranks from the communicator prior to a blocking collective call and the other ranks simply waited and resumed afterward once the delayed rank were ready.

Perfect, in that case we are good to go. Thanks for confirming @jnke2016 .